### ISSUE TITLE
Performance problem on listings and widgets with Price box

### ISSUE THREAD

**Issue by ilnytskyi** [magento/magento2#26850]
### Preconditions (*)

1. 2.4-develop or any previous version
2.

Usage of `\Magento\Catalog\Block\Product\ListProduct::getProductPrice`
in `app/code/Magento/Catalog/view/frontend/templates/product/list.phtml` or product boxes in widgets affects listing performance.
It calculates price for products again even if the values received from collection.
On the screenshot you can see that product items have price values.
![Selection_047](https://user-images.githubusercontent.com/3192770/74373133-23fabc00-4ddc-11ea-90b3-1a9586a29161.png)

Then when a product is rendered, magento starts to resolve the price
![Selection_048](https://user-images.githubusercontent.com/3192770/74373484-dc286480-4ddc-11ea-8bf9-7a82816aa457.png)

It does these steps:
- Looks for price renderer
- loads linked products for composite products
- Resolve price type (it already available from collection)

Because the prices already available from the collection it's enough to render formatted price value (amount and currency) [e.g. 10.00 $]. But magento loads prices again does comparations and then renders it. Even when sort by price used.

### Steps to reproduce (*)

1. Render price box from the product collection on listing or search pages
2. Use simple or composite products
3. See that price boxes consumes a lot of time for price rendering by calling 
`\Magento\Catalog\Block\Product\ListProduct::getProductPrice` in the loop

### Expected result (*)

1. Prices are rendered faster
2. Redundant loads and price comparations in `ListProduct::getProductPrice` are eliminated
3. `ListProduct::getProductPrice` just formats given price float values. It does not resolve prices again
4. Moreover it does not load linked/child products

### Actual result (*)

1. Prices are rendered slowly
2. `ListProduct::getProductPrice` loads extra price data and performs comparations
3. `ListProduct::getProductPrice` does not use data already available from collection, loads prices again
4. `ListProduct::getProductPrice` loads child/linked products and performs redundant price resolve.

**Comment by mrtuvn** [comment]
Pricing is still hard parts in magento 2 and undoubtedly need refactors and improvements

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Delta 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-31591` were created

**Issue Available**: @engcom-Delta, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by verheesj** [comment]
Isn't this because the product in question is a configurable product?

**Comment by ilnytskyi** [comment]
@verheesj 
yes, but the problem is in its price renderer implementation

**Comment by VladimirZaets** [comment]
**Action items to fix:**

- Verify that recalculated data is the same as the data from the index table ( collection )
- If data the same, refactor implementation to use data from the index (collection)
- If not, investigate the reason why data is different

**Action items to testing:**
- Test fix with all types of price

**Comment by mrtuvn** [comment]
This one still valid today. Hope someone can take a look at it

**Comment by Nuranto** [comment]
Hope we'll have a fix for this in 2.4.5 !

**Comment by Nuranto** [comment]
still valid on 2.4.5

**Comment by MTheProgrammer** [comment]
Same here, prices are loaded separately for every product of the configuration.

**Comment by lukutpg** [comment]
I had also performance issue for configurable product which have almost 2000 simple products (Magento CE  2.4.4).
For me the most slow was function `\Magento\InventoryConfigurableProduct\Plugin\Model\Product\Type\Configurable\IsSalableOptionPlugin::filterProducts`, so I just overridden it with follow code:
```diff
--- vendor/magento/module-inventory-configurable-product/Plugin/Model/Product/Type/Configurable/IsSalableOptionPlugin.php   
+++ vendor/magento/module-inventory-configurable-product/Plugin/Model/Product/Type/Configurable/IsSalableOptionPlugin.php
@@ -106,15 +106,24 @@
      * @param array $isSalableResults
      * @return void
      */
-    private function filterProducts(array $products, array $isSalableResults) : void
+    private function filterProducts(array &$products, array $isSalableResults) : void
     {
+        $isSalableResultsTest = [];
+        foreach ($isSalableResults as $result) {
+            if (!$result->isSalable()) {
+                $isSalableResultsTest[$result->getSku()] = $result->isSalable();
+            }
+        }
+
+        if (empty($isSalableResultsTest)) {
+            return;
+        }
+
         foreach ($products as $key => $product) {
-            foreach ($isSalableResults as $result) {
-                if ($result->getSku() === $product->getSku() && !$result->isSalable()) {
-                    $product->setIsSalable(0);
-                    if (!$this->stockConfiguration->isShowOutOfStock()) {
-                        unset($products[$key]);
-                    }
+            if (array_key_exists($product->getSku(), $isSalableResultsTest)) {
+                $product->setIsSalable(0);
+                if (!$this->stockConfiguration->isShowOutOfStock()) {
+                    unset($products[$key]);
                 }
             }
         }
```

**Comment by MTheProgrammer** [comment]
Exactly, MSI is querying every stock for every product. Another newralgic point is swatches renderer which loads all options for every configurable product on the listing.

**Comment by qsolutions-pl** [comment]
Latest version of M2 has the same issue.

**Comment by hubertus2017** [comment]
For google shopping we there upload each simple products of configured products with their prices and individual links to preselect the option, e.g.

https........html?ps=MTU4PTIxOCYxNTk9MjIy#158=218&159=222

When loading the page, for the first second it displays the "as From" price from the lowest price of another simple product of the configurable product. Later (3 second?) it renders the correct price of the simple product. But that's too late for the google shopping crawler. He takes the first displayed "from price", which is incorrect. this leads to penalties at google and wrong ads.

I looked into final_price.phtml files but it did not help. i believe the Magento_Catalog/js/price-box.js must be changed?

Either to get rid of the display of the "from price" and/or how to speed up the display of the correct price?

Magento ver. 2.4.4

**Comment by ilnytskyi** [comment]
@jlenzstaysafe try this module https://github.com/EcomDev/magento2-product-preloader
As claimed it fixes `Price of configurable product on product list` can be some sort of workaround

**Comment by qsolutions-pl** [comment]
@ilnytskyi  @jlenzstaysafe  this module is just a one part, it still requires to create a custom module and optimization of SQL queries. But yes, using it you drop PLP page to under 700ms load time

**Comment by hryvinskyi** [comment]
Any news?
